rust/monovertex/src/forwarder.rs

```diff
@@ -205,67 +210,62 @@ impl Forwarder {
+                    // create a map of id to result, since there is no strict requirement
+                    // for the udsink to return the results in the same order as the requests
+                    let result_map: HashMap<_, _> = response
                         .results
                         .iter()
-                        .filter(|result| result.status == proto::Status::Failure as i32)
-                        .map(|result| {
-                            messages
-                                .iter()
-                                .find(|msg| msg.id == result.id)
-                                .unwrap()
-                                .clone()
-                        })
-                        .collect::<Vec<_>>();
-
-                    if messages.is_empty() {
-                        break;
-                    } else {
-                        error_map.clear();
-                        for result in response.results {
-                            if result.status == proto::Status::Failure as i32 {
-                                *error_map.entry(result.err_msg).or_insert(0) += 1;
-                            }
+                        .map(|result| (result.id.clone(), result))
+                        .collect();
+
+                    // drain all the messages that were successfully written
+                    // and keep only the failed messages to send again
+                    messages_to_send.retain(|msg| {
+                        if let Some(result) = result_map.get(&msg.id) {
+                            return if result.status == proto::Status::Success as i32 {
+                                false
+                            } else if result.status == proto::Status::Fallback as i32 {
+                                fallback_msgs.push(msg.clone()); // add to fallback messages
+                                false
+                            } else {
+                                *error_map.entry(result.err_msg.clone()).or_insert(0) += 1;
+                                true
+                            };
                         }
+                        false
+                    });
 
-                        warn!(
-                            "Retry attempt {} due to retryable error. Errors: {:?}",
-                            attempts, error_map
-                        );
-                        sleep(tokio::time::Duration::from_millis(
-                            config().sink_retry_interval_in_ms as u64,
-                        ))
-                        .await;
+                    // if all messages are successfully written, break the loop
+                    if messages_to_send.is_empty() {
+                        break;
                     }
+
+                    warn!(
+                        "Retry attempt {} due to retryable error. Errors: {:?}",
+                        attempts, error_map
+                    );
+                    error_map.clear();
```

Comment:
if we clear here, how will we return it in the `Err` ?

Reply:
updated to handle that case.